apps/files_versions/lib/Versions/LegacyVersionsBackend.php

```diff
@@ -196,6 +197,21 @@ public function read(IVersion $version) {
 
 	public function getVersionFile(IUser $user, FileInfo $sourceFile, $revision): File {
 		$userFolder = $this->rootFolder->getUserFolder($user->getUID());
+		$ownerUserId = $sourceFile->getOwner()?->getUID();
+		$storage = $sourceFile->getStorage();
+
+		if ($storage->instanceOfStorage(ISharedStorage::class) && $ownerUserId) {
+			/** @var SharedStorage $storage */
+			// Shared files have their versions in the owners root folder
+			$userFolder = $this->rootFolder->getUserFolder($ownerUserId);
+			$user = $this->userManager->get($ownerUserId);
+			$ownerPathInStorage = $sourceFile->getInternalPath();
+			$sourceFile = $storage->getShare()->getNode();
+			if ($sourceFile instanceof Folder) {
+				$sourceFile = $sourceFile->get($ownerPathInStorage);
+			}
```

Comment:
I don't get this part of the code, why the distinction?

Reply:
Ok, but now I do not understand why we are overriding `$sourceFile`.

If we want to have a `Node` object relative to the owner's folder, would that work?

```suggestion
			$sourceFile = $userStorage->get($sourceFile->getId());
```

Reply:
Getting the file by id might be slower here as it would require setting up mountpoints until the file with id is found in the provided user home. Getting by path can take more shortcuts there by only setting up mountpoints in the path that is requested.